chainstate/src/detail/spend_cache/mod.rs

```diff
@@ -498,6 +602,191 @@ impl<'a, S: BlockchainStorageRead> CachedInputs<'a, S> {
         Ok(())
     }
 
+    fn check_connected_transfer_data(
+        &self,
+        block_id: Id<Block>,
+        tx: &Transaction,
+        token_id: &TokenId,
+        amount: &Amount,
+    ) -> Result<(), TokensError> {
+        // Collect token inputs
+        let total_value_tokens = self
+            .calculate_assets_total_inputs(tx.inputs())
+            .map_err(|_err| TokensError::NoTokenInInputs(tx.get_id(), block_id))?;
+
+        // Is token exist in inputs?
+        let (_, origin_amount) = total_value_tokens
```

Comment:
This implementation is incorrect.

To prove this, create a test that has one input of token X with value 100. Then create two or more outputs with value 100 + 100 + 100 + 100... You basically can print as much tokens as you want with this bug by having more outputs that spend the same input.

The mistake here is that you're not removing the tokens you marked as spent from the map. Since we're doing things functionally and we dislike using mutable variables, the better way is to create a map with the "spent tokens", similar to the total_value_tokens map, but then every output you "consume" adds to that map (here in this line of the code). Then, your comparison should be with the value of the token in the map total_value_tokens minus the one in the map that have the "consumed" tokens.

Reply:
A simpler solution is perhaps to just create another map for the total in outputs, and compare.

Reply:
Fixed!